### ISSUE TITLE
implement per-directory configuration

### ISSUE THREAD

**Issue by vgough** [vgough/encfs#5]
This could be used to provide per-directory keys or IVs, replacing directory IV chaining and the horrible recursive rename logic.

**Comment by benrubson** [comment]
As I said in #79, I do not use _Filename Initialization Vector Chaining_ because I have directories with a huge amout of subdirectories and files, so I do not want to be impacted by the significant performance impact which directory renames involve.

Could we think about a solution to give each object's name its own initialization vector but not based on its path ?
This would help with the performance impact.

As you said @vgough in #79 :
> Yes, I would like to replace the full chaining with a per-directory IV, so that renaming a directory does not require changing contents of the directory. AFAIK, nobody is working on it yet, it's just a wishlist item.

Thank you !

Ben

**Comment by rfjakob** [comment]
If you want this feature, you should probably take a look at [gocryptfs](https://github.com/rfjakob/gocryptfs). It uses a file "gocryptfs.diriv" in each directory. Disclaimer: I am the author.

**Comment by rfjakob** [comment]
That said, I'm not opposed to adding it to EncFS, too. But it's a lot of work in a sometimes difficult to work with code base. And a huge user base where you have to be very careful not to introduce behavoir changes.

You can check https://github.com/rfjakob/gocryptfs/issues/49 for my reasoning for starting afresh instead.

**Comment by benrubson** [comment]
> I'm not opposed to adding it to EncFS

What is the minimum required size for an IV ?
What about finding a way to put it at the beginning (or at the end) of the directory name itself ?
We would then avoid the _one-file-per-directory_ penalty 👍 
I'm not so found of having to use additional files to manage user-data encryption :|

**Comment by rfjakob** [comment]
No, don't want to have random data in the file or directory name.

If you did, you'd have to read the whole directory to open a single file.

**Comment by rfjakob** [comment]
Size should be >= 16 bytes.

**Comment by benrubson** [comment]
> If you did, you'd have to read the whole directory to open a single file.

Let's imagine we encrypt `mydirectory` to `<16bytesIV><encrypt(mydirectory)>`.
The `<16bytesIV>` would then be used to encrypt filenames under `mydirectory`, it won't be used to encrypt `mydirectory` name itself (`mydirectory` will use the IV from its parent).
So, if we want to decrypt `<encrypt(myfile)>` which is a file under `mydirectory`, we only have to read this file, as we already have its name IV (`<16bytesIV>`), because we had to enter `mydirectory`.
You see ?

**Comment by rfjakob** [comment]
How do you open mydirectory? Part of its name is random!

**Comment by benrubson** [comment]
Damn ! You're right ! For sure of course we won't be able to directly open `mydirectory`... Sorry...

We could derivate IV from the inode number, but everything would be broken if we move / duplicate the encrypted folder somewhere else.

Perhaps we could store the IV in an  extended attribute to the directory.
Extended attributes are metadata, we would then not have a data performance penalty.

We could also as you said (as you did with gocryptfs) use a file (but at some performance cost).

I don't see other solutions.
Did you find others working on this for gocryptfs ?

**Comment by rfjakob** [comment]
I considered extended attributes, but ultimatetely went for the plain file
for gocryptfs, because it's quite easy to lose extended attributes if you
are not very careful.
Think dropbox, fat32 usb sticks, zip files or even  a .tar.gz where you
forgot to set the right options.

**Comment by benrubson** [comment]
Yep, could be dangerous.
Possibly for expert users.
